src/QirRuntime/lib/QIR/utils.cpp

```diff
@@ -0,0 +1,54 @@
+#include <assert.h>
+#include <stdexcept>
+#include <memory>
+#include <unordered_set>
+#include <iostream>
+
+#include "__quantum__rt.hpp"
+
+#include "qirTypes.hpp"
+
+std::unordered_set<char*>& UseMemoryTracker()
+{
+    static std::unordered_set<char*> memoryTracker;
+    return memoryTracker;
+}
+
+extern "C"
+{
+    // Allocate a block of memory on the heap.
+    char* quantum__rt__heap_alloc(int size) // NOLINT
+    {
+        char* buffer = new char[size];
+        #ifndef NDEBUG
+            UseMemoryTracker().insert(buffer);
+        #endif
+        return buffer;
+    }
+
+    // Release a block of allocated heap memory.
+    void quantum__rt__heap_free(char* buffer) // NOLINT
+    {
+        #ifndef NDEBUG
+            auto iter = UseMemoryTracker().find(buffer);
+            assert(iter != UseMemoryTracker().end());
+            UseMemoryTracker().erase(iter);
+        #endif
+        delete[] buffer;
+    }
+
+    // Fail the computation with the given error message.
+    void quantum__rt__fail(QirString* msg) // NOLINT
+    {
+        std::string str = msg->str;
+        quantum__rt__string_unreference(msg);
+        throw std::runtime_error(str);
+    }
+
+    // Include the given message in the computation's execution log or equivalent.
+    // TODO: should we allow the user to register their own output?
+    void quantum__rt__message(QirString* msg) // NOLINT

```

Comment:
Yes, a very valid point and I was on the fence about it. The RT defines all APIs in a target-agnostic way. It happens, that when targeting simulators, we delegate some of the APIs to the simulators (e.g. qubit allocation) because it makes sense for these targets. We could pass message call to the simulator as well. However, we don't have to (as this implementation shows) and I'd like to keep the IQuantumApi as thin as possible. On the other hand, I agree that dumping to std::cout might not be suitable for all situations... We should probably have a notion of execution context that incorporates the target simulator and configurations for things like message/fail (as I'm also not convinced that throwing from `fail` is always the correct choice). I'd suggest we leave this implementation in for now and review it later when we have better understanding of the ways QIR is produced/consumed and the requirements that places on the runtime.